unison-cli/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -2899,24 +2865,107 @@ loadPropagateDiffDefaultPatch inputDescription maybeDest0 dest = do
+  let endangeredDeletions = List.filter (\m -> not $ null m || Map.foldr (\s b -> null s || b ) False m ) endangered
+  if null endangeredDeletions then do
+    let deleteTypesTerms =
+          splitsNames >>= (\(split, _, types, terms) ->
+            (map (BranchUtil.makeDeleteTypeName split) . Set.toList $ types) ++
+            (map (BranchUtil.makeDeleteTermName split) . Set.toList $ terms )
+          )
+    before <- Cli.getRootBranch0
+    description <- inputDescription inputs
+    Cli.stepManyAt description deleteTypesTerms
+    case doutput of
+      DeleteOutput'Diff -> do
+        after <- Cli.getRootBranch0
+        (ppe, diff) <- diffHelper before after
+        Cli.respondNumbered (ShowDiffAfterDeleteDefinitions ppe diff)
+      DeleteOutput'NoDiff -> do
+        Cli.respond Success
+  else do
+      ppeDecl <- currentPrettyPrintEnvDecl Backend.Within
+      let combineRefs = List.foldl (Map.unionWith NESet.union) Map.empty endangeredDeletions
+      Cli.respondNumbered (CantDeleteDefinitions ppeDecl combineRefs)
+
 -- | Goal: When deleting, we might be removing the last name of a given definition (i.e. the
 -- definition is going "extinct"). In this case we may wish to take some action or warn the
 -- user about these "endangered" definitions which would now contain unnamed references.
 getEndangeredDependents ::
-  -- | Which names we want to delete
+  -- | Single target for deletion
   Names ->
+  -- | All names we want to delete (including the target)
+  Set LabeledDependency ->
   -- | All names from the root branch
   Names ->
   -- | map from references going extinct to the set of endangered dependents
   Sqlite.Transaction (Map LabeledDependency (NESet LabeledDependency))
-getEndangeredDependents namesToDelete rootNames = do
+getEndangeredDependents targetNamesToDelete allTermsToDelete rootNames = do
+  -- names of terms left over after target deletion
   let remainingNames :: Names
-      remainingNames = rootNames `Names.difference` namesToDelete
-      refsToDelete, remainingRefs, extinct :: Set LabeledDependency
-      refsToDelete = Names.labeledReferences namesToDelete
-      remainingRefs = Names.labeledReferences remainingNames -- left over after delete
-      extinct = refsToDelete `Set.difference` remainingRefs -- deleting and not left over
-      accumulateDependents :: LabeledDependency -> Sqlite.Transaction (Map LabeledDependency (Set LabeledDependency))
+      remainingNames = rootNames `Names.difference` targetNamesToDelete
+  let refsToDelete :: Set LabeledDependency
+      refsToDelete = Names.labeledReferences targetNamesToDelete
+  -- remove the target from the set of names to delete
+  let allOtherNamesToDelete :: Set LabeledDependency
+      allOtherNamesToDelete = Set.difference allTermsToDelete refsToDelete
+  -- left over after deleting target
+  let remainingRefs :: Set LabeledDependency
+      remainingRefs =  Names.labeledReferences remainingNames
```

Comment:
Oof, this seems really expensive. This is another reason we'd really like to limit deletion checks to the current path if possible (but if that's out of scope we can make a new issue for that).

Many of these steps seem to be mapping over ALL the names in the whole root namespace haha. Laziness will save us a little work, but it's hard to say how much.